activemodel/lib/active_model/errors.rb

```diff
@@ -205,22 +205,56 @@ def attribute_names
-    # if the json object should contain full messages or not (false by default).
-    #
-    #   person.errors.as_json                      # => {:name=>["cannot be nil"]}
-    #   person.errors.as_json(full_messages: true) # => {:name=>["name cannot be nil"]}
-    def as_json(options = nil)
-      to_hash(options && options[:full_messages])
+    # object. Use the <tt>message_method: <method_mame></tt> argument
+    # with the values <tt>:message</tt>, <tt>:custom_message</tt> and <tt>:full_message</tt>
+    # to format the message with the desired method. By default it'll be <tt>:message</tt>.
+    # See `error.rb` for the definition of each method.
+    #
+    #   person.errors.as_json                                # => {:name=>["cannot be nil"]}
+    #   person.errors.as_json(message_method: :full_message) # => {:name=>["name cannot be nil"]}
+    def as_json(**options)
+      # to_hash(options && options[:full_messages])
+      #
+      # # message => { :name => ["cannot be nil"] }
+      # # full_message => { :name => ["Name cannot be nil"] }
+      # # custom_message => { :name => ["Whoops! You need to provide a value"]}
+      return to_hash if options.nil?
+
+      if options.keys.include?(:full_messages)
+        ActiveSupport::Deprecation.warn(
+          "using the full_messages boolean positional argument is deprecated" \
+          "and will be unsupported in a future version of Rails." \
+          "Use the method_type keyword argument with either :message, :full_message, or :custom_message"
+        )
+        to_hash(options && options[:full_messages])
+      else
+        to_hash(**options)
+      end
+    end
+
+    def to_json(**options)
+      as_json(**options).to_json
     end
 
-    # Returns a Hash of attributes with their error messages. If +full_messages+
-    # is +true+, it will contain full messages (see +full_message+).
+    # Returns a Hash of attributes with their error messages.
+    # Use the <tt>message_method: <method_mame></tt> argument
+    # with the values <tt>:message</tt>, <tt>:custom_message</tt> and <tt>:full_message</tt>
+    # to format the message with the desired method. By default it'll be <tt>:message</tt>.
+    # See `error.rb` for the definition of each method.
     #
     #   person.errors.to_hash       # => {:name=>["cannot be nil"]}
     #   person.errors.to_hash(true) # => {:name=>["name cannot be nil"]}
-    def to_hash(full_messages = false)
-      message_method = full_messages ? :full_message : :message
+    def to_hash(full_messages = false, **options)
+      if options[:message_method]
+        message_method = options[:message_method]
+      else
+        ActiveSupport::Deprecation.warn(
+          "using the full_messages boolean positional argument is deprecated. " \
+          "Use the method_type keyword argument with :message, :full_message, or :custom_message"
+        )
+        message_method = full_messages ? :full_message : :message
+      end
+
```

Comment:
If `:message_method` is the only option we can make it more explicit with named arguments:

```suggestion
    def to_hash(full_messages = false, message_method: nil)
      unless message_method
        ActiveSupport::Deprecation.warn(
          "using the full_messages boolean positional argument is deprecated. " \
          "Use the method_type keyword argument with :message, :full_message, or :custom_message"
        )
        message_method = full_messages ? :full_message : :message
      end
      
```